lib/aerospike/cluster.rb

```diff
@@ -229,84 +218,96 @@ def launch_tend_thread
       end
     end
 
+    # Check health of all nodes in cluster
     def tend
-      nodes = self.nodes
+      was_changed = refresh_nodes
+
+      return unless was_changed
+
+      update_cluster_features
+      notify_cluster_config_changed
+      # only log the tend finish IF the number of nodes has been changed.
+      # This prevents spamming the log on every tend interval
+      log_tend_stats(nodes)
+    end
+
+    # Refresh status of all nodes in cluster. Adds new nodes and/or removes
+    # unhealty ones
+    def refresh_nodes
       cluster_config_changed = false
 
-      # All node additions/deletions are performed in tend thread.
-      # If active nodes don't exist, seed cluster.
+      nodes = self.nodes
       if nodes.empty?
-        Aerospike.logger.info("No connections available; seeding...")
         seed_nodes
         cluster_config_changed = true
-
-        # refresh nodes list after seeding
         nodes = self.nodes
       end
 
-      # Refresh all known nodes.
-      friend_list = []
-      refresh_count = 0
+      peers = Peers.new
 
-      # Clear node reference counts.
+      # Clear node reference count
       nodes.each do |node|
-        node.reference_count.value = 0
-        node.responded.value = false
+        node.reset_reference_count!
+        node.reset_responded!
+        node.partition_generation.reset_changed!
+        # Using peers is default true
+        peers.use_peers = false unless node.supports_feature?('peers')
```

Comment:
This could be pulled out of the loop:

    peers.use_peers = nodes.all? { |node| node.supports_feature?('peers') }

(Can also move this into a separate method, e.g. `Cluster#supports_peers_protocol?`)

Then the rest of the "reset_*" calls can be moved into `Node::Refresh::Info` or a new service method (`Node::Refresh::Reset`?). The cluster model shouldn't really need to be aware of all the per-node book-keeping.